service/history/conflictResolverV2.go

```diff
@@ -0,0 +1,186 @@
+	}
+	//TODO: use LCA on the msBuilder.getVersionHistories() vs localVersionHistories
+	//r.resolveVersionHistoryConflict(localVersionHistories, msBuilder.GetVersionHistories())
+	return msBuilder, err
+}
+
+func (r *conflictResolverV2Impl) resolveVersionHistoryConflict(local persistence.VersionHistories, remote persistence.VersionHistories) error {
+
+	for _, newHistory := range remote.GetHistories() {
+		commonEventId, history, err := local.FindLowestCommonVersionHistory(newHistory)
+		if err != nil {
+			return err
+		}
+
+		if err := local.AddHistory(commonEventId, history, newHistory); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *conflictResolverV2Impl) getHistoryBatch(domainID string, execution shared.WorkflowExecution, firstEventID,
+	nextEventID int64, nextPageToken []byte, eventStoreVersion int32, branchToken []byte) collection.PaginationFn {
+	return func(paginationToken []byte) ([]interface{}, []byte, error) {
+		var paginateItems []interface{}
+		_, historyBatches, token, _, err := PaginateHistory(
+			r.historyMgr,
+			r.historyV2Mgr,
+			nil,
+			r.logger,
+			true,
+			domainID,
+			*execution.WorkflowId,
+			*execution.RunId,
+			firstEventID,
+			nextEventID,
+			nextPageToken,
+			eventStoreVersion,
+			branchToken,
+			defaultPageSize,
+			common.IntPtr(r.shard.GetShardID()))
+		for _, history := range historyBatches {
+			paginateItems = append(paginateItems, history)
+		}
+		return paginateItems, token, err
+	}
+}
+
+func (r *conflictResolverV2Impl) initializeBuilders(firstEvent *shared.HistoryEvent) (*mutableStateBuilder, *stateBuilderImpl) {
+	resetMutableStateBuilder := newMutableStateBuilderWithReplicationState(
+		r.clusterMetadata.GetCurrentClusterName(),
+		r.shard,
+		r.shard.GetEventsCache(),
+		r.logger,
+		firstEvent.GetVersion(),
+	)
+	resetMutableStateBuilder.executionInfo.EventStoreVersion = eventStoreVersion
+	sBuilder := newStateBuilder(r.shard, resetMutableStateBuilder, r.logger)
+	return resetMutableStateBuilder, sBuilder
+}
+
+func (r *conflictResolverV2Impl) resetMutableState(msBuilder *mutableStateBuilder,
+	execution *persistence.WorkflowExecutionInfo,
+	lastEvent *shared.HistoryEvent,
+	replayEventID int64,
+	prevRunID string) (mutableState, error) {
+	// reset branchToken to the original one(it has been set to a wrong branchToken in applyEvents for startEvent)
+	msBuilder.executionInfo.BranchToken = msBuilder.GetCurrentBranch()
+	// similarly, in case of resetWF, the runID in startEvent is incorrect
+	msBuilder.executionInfo.RunID = execution.RunID
+	msBuilder.executionInfo.StartTimestamp = execution.StartTimestamp
+	// the last updated time is not important here, since this should be updated with event time afterwards
+	msBuilder.executionInfo.LastUpdatedTimestamp = execution.StartTimestamp
+	sourceCluster := r.clusterMetadata.ClusterNameForFailoverVersion(lastEvent.GetVersion())
+	msBuilder.UpdateReplicationStateLastEventID(sourceCluster, lastEvent.GetVersion(), replayEventID)
```

Comment:
the following 2 lines are handled by state builder
```
sourceCluster := r.clusterMetadata.ClusterNameForFailoverVersion(lastEvent.GetVersion())
msBuilder.UpdateReplicationStateLastEventID(sourceCluster, lastEvent.GetVersion(), replayEventID)
```

Reply:
remvoed

Reply:
removed